boa_gc/src/cell.rs

```diff
@@ -0,0 +1,556 @@
+    }
+}
+
+impl<'a, T: Trace + ?Sized, U: ?Sized> DerefMut for GcCellRefMut<'a, T, U> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut U {
+        self.value
+    }
+}
+
+impl<'a, T: Trace + ?Sized, U: ?Sized> Drop for GcCellRefMut<'a, T, U> {
+    #[inline]
+    fn drop(&mut self) {
+        debug_assert!(self.gc_cell.flags.get().borrowed() == BorrowState::Writing);
+        // Restore the rooted state of the GcCell's contents to the state of the GcCell.
+        // During the lifetime of the GcCellRefMut, the GcCell's contents are rooted.
+        if !self.gc_cell.flags.get().rooted() {
+            // SAFETY: If `GcCell` is no longer rooted, then unroot it. This should be safe
+            // as the internal `GcBox` should be guaranteed to have at least 1 root.
+            unsafe {
+                (*self.gc_cell.cell.get()).unroot();
+            }
+        }
+        self.gc_cell
+            .flags
+            .set(self.gc_cell.flags.get().set_unused());
+    }
+}
+
+impl<'a, T: Trace + ?Sized, U: Debug + ?Sized> Debug for GcCellRefMut<'a, T, U> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Debug::fmt(&**self, f)
+    }
+}
+
+impl<'a, T: Trace + ?Sized, U: Display + ?Sized> Display for GcCellRefMut<'a, T, U> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Display::fmt(&**self, f)
+    }
+}
+
+// SAFETY: GcCell<T> tracks it's `BorrowState` is `Writing`
+unsafe impl<T: ?Sized + Send> Send for GcCell<T> {}
+
+impl<T: Trace + Clone> Clone for GcCell<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        Self::new(self.borrow().clone())
+    }
+}
+
+impl<T: Trace + Default> Default for GcCell<T> {
+    #[inline]
+    fn default() -> Self {
+        Self::new(Default::default())
+    }
+}
+
+#[allow(clippy::inline_always)]
+impl<T: Trace + ?Sized + PartialEq> PartialEq for GcCell<T> {
+    #[inline(always)]
+    fn eq(&self, other: &Self) -> bool {
+        *self.borrow() == *other.borrow()
+    }
+}
+
+impl<T: Trace + ?Sized + Eq> Eq for GcCell<T> {}
+
+#[allow(clippy::inline_always)]
+impl<T: Trace + ?Sized + PartialOrd> PartialOrd for GcCell<T> {
+    #[inline(always)]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        (*self.borrow()).partial_cmp(&*other.borrow())
+    }
+
+    #[inline(always)]
+    fn lt(&self, other: &Self) -> bool {
+        *self.borrow() < *other.borrow()
+    }
+
+    #[inline(always)]
+    fn le(&self, other: &Self) -> bool {
+        *self.borrow() <= *other.borrow()
+    }
+
+    #[inline(always)]
+    fn gt(&self, other: &Self) -> bool {
+        *self.borrow() > *other.borrow()
+    }
+
+    #[inline(always)]
+    fn ge(&self, other: &Self) -> bool {
+        *self.borrow() >= *other.borrow()
+    }
+}
+
+impl<T: Trace + ?Sized + Ord> Ord for GcCell<T> {
+    #[inline]
+    #[allow(clippy::use_self)]
+    fn cmp(&self, other: &GcCell<T>) -> Ordering {
+        (*self.borrow()).cmp(&*other.borrow())
+    }
+}
```

Comment:
Any reason against this?
```suggestion
    fn cmp(&self, other: &Self) -> Ordering {
        (*self.borrow()).cmp(&*other.borrow())
    }
}
```

Reply:
I just thought it might lack a little bit of clarity to go be comparing &self vs &Self

Reply:
Looking at the documentation for [Ord](https://doc.rust-lang.org/std/cmp/trait.Ord.html#tymethod.cmp) this is the normal signature, I think that is why clippy recommends this.